### PR TITLE
Feat audio streaming

### PR DESCRIPTION
**Pull request by EzraEllette**
## description
This PR changes the way audio is gathered into a stream that can be subscribed to. This is the foundation of features such as realtime transcription.

related issue: #521 

## type of change
- [ ] bug fix
- [x] new feature
- [ ] breaking change
- [ ] documentation update

## how to test
1. Verify audio transcription is working as expected.

## checklist
- [x] i have read the [CONTRIBUTING.md](https://github.com/mediar-ai/screenpipe/blob/main/CONTRIBUTING.md) file 
- [x] i have added the custom cursor AI prompt to my settings as mentioned in CONTRIBUTING.md and used to write this PR
- [x] my code follows the project's style guidelines
- [x] i have performed a self-review of my code
- [x] i have updated the documentation if necessary
- [x] my changes generate no new warnings
- [x] i have added tests that prove my fix is effective or that my feature works
- [x] all tests pass locally with my changes

### CODE REVIEW

File: screenpipe-audio/src/core.rs [hunk 1]
```diff
@@ -159,176 +160,57 @@ async fn get_device_and_config(
...
-
-    debug!("sending audio to audio model");
-    if let Err(e) = whisper_sender.send(AudioInput {
-        data: Arc::new(audio_data),
-        device: audio_device.clone(),
-        sample_rate,
-        channels,
-    }) {
-        error!("failed to send audio to audio model: {}", e);
+            // Reset collected audio to the last two seconds of recorded audio
```
Comment (louis030195) [inline screenpipe-audio/src/core.rs:205]: why do we need overlap if streaming audio?

unrelated: i was wondering how we can improve the use case where you have a speaker and the mic and system audio basically duplicate the transcription (probably in this case should only use the system audio), not a priority

**Comment by louis030195** [comment]
great work, will test shortly and merge today if it make sense 

if anyone on windows or linux can test too would appreciate! (i will test on windows too but it's high friction for me)

File: screenpipe-audio/src/core.rs (hunk 1 above)
Comment (EzraEllette) [reply screenpipe-audio/src/core.rs:205]: The chunk my end in the middle of a word. realtime VAD would help with this.

**Comment by louis030195** [comment]
can i merge?

**Comment by EzraEllette** [comment]
Did you test on windows?

**Comment by louis030195** [comment]
will test on windows now then merge if it works

**Comment by louis030195** [comment]
> will test on windows now then merge if it works

update: cannot test, trying to fix build first, also my windows is just slow as fuck takes me 40 min to run a script

**Comment by EzraEllette** [comment]
My windows laptop is fast but I left my charger 🙃. I'm trying to get ahold of one but stores are closed right now.

**Comment by louis030195** [comment]
@tribhuwan-kumar any chance you can test this on windows?

**Comment by tribhuwan-kumar** [comment]
> @tribhuwan-kumar any chance you can test this on windows?

sure, let me test it real quick

**Comment by EzraEllette** [comment]
image

**Comment by EzraEllette** [comment]
@louis030195

**Comment by tribhuwan-kumar** [comment]
@louis030195 everything is fine, there are not any unexpected behaviors. 
[test.log](https://github.com/user-attachments/files/17482353/test.log)
[screenpipe-app.log](https://github.com/user-attachments/files/17482354/screenpipe-app.log)
[screenpipe.2024-10-22.log](https://github.com/user-attachments/files/17482362/screenpipe.2024-10-22.log)

**Pull request merged**